### DISCUSSION TITLE
Login via token obtained from web firebase auth sdk

### DISCUSSION THREAD

**Discussion by devth** [invertase/react-native-firebase#5800] [Q&A]
I have a hybrid RN app where much of the UI is embedded in a WebView and some of it is in RN screens. I'm using `@react-native-firebase/auth` to interact with Firebase from the RN screens, but the user logs in via the web app in the WebView.

How can I use the tokens from the authenticated user in the web view to set the authenticated user in RN? In other words: I want to sync login state between RN and web.

**Comment by mikehardy** [comment]
I've never contemplated that use case, so I'm not sure at all, sorry. I'm not aware of any APIs that let you set an externally fetched firebase user auth token into the native SDKs. If I understand correctly they assume you will use native auth methods to authenticate, and they will handle token creation/exchange and give the token to you, with no possibility for the reverse path

**Reply by devth** [reply]
Thanks for the quick reply @mikehardy. I'll keep working on a solution or workaround and post back here if I figure it out!

**Reply by devth** [reply]
I did find [this StackOverflow post](https://stackoverflow.com/questions/58368242/how-to-pass-firebase-auth-token-to-webview-and-register-for-notifications-on-and) on the opposite direction, i.e. login with RN then pass the token to the webview to be used with [`firebase.auth.signInWithCredential`](https://firebase.google.com/docs/reference/js/v8/firebase.auth.Auth#signinwithcredential), which takes an [AuthCredential](https://firebase.google.com/docs/reference/js/v8/firebase.auth.AuthCredential).

**Reply by mikehardy** [reply]
Interesting - we do have the signInWithCredential method in react-native-firebase as well, if that's useful, perhaps it could go either direction. I'm just not sure how to pass things back and forth with a webview, there are `@JavascriptInterface` possibilities if you're a mixed app, you could make a small react-native native module that can accept (or pass) credentials by maintaining state, with methods marked as javascript interfaces, then in your webview you can call those methods to pass them in, or in react-native you can call the `@ReactMethods` to get them. Just brainstorming, I won't be able to actually help sorry

**Reply by devth** [reply]
Thanks, I'll see what I can figure out!

**Reply by hronnro** [reply]
Hey @devth 👋 

I have the same idea in mind but I haven't even tried implementing it yet so I'm happy I found this thread! 🙏

Did you manage to solve this in your app?

**Reply by devth** [reply]
Yep, I got this working!

Here's what I run on the RN side:

```typescript
  const [user, setUser] = useGlobalState("user"); // this uses react-hooks-global-state lib - you can store state however you want

  const authenticate = async (customToken: string) => {
    analytics().logEvent("login_attempt");

    try {
      const authResult = await auth().signInWithCustomToken(customToken);
      if (authResult.user) {
        analytics().logEvent("login");
        setUser(authResult.user);
      }
      return authResult;
    } catch (error) {
      console.error("Error authenticating", error);
      crashlytics().log(`error authenticating: ${error.message}`);
      crashlytics().recordError(error);
    }
  };
```

on the web side, that custom token is generated using the admin sdk like:

```typescript
        const customToken = await admin.auth().createCustomToken(user.uid);
```

I then use the RN bridge to communicate between the webview and the RN wrapper.

**Reply by hronnro** [reply]
yeesss TYSM!! 🙌 (this will literally save me so much work lol)

**Reply by devth** [reply]
Glad I could help! 

I can post more details about the React Native webview bridge comm if it helps too.

**Reply by CodeDeyya** [reply]
hi does this work on client side or just the server side ? Ive heard it does not work on client side. I am trying to bridge to a react application.
